xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java

```diff
@@ -3446,11 +3457,150 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+                    parameter.getKey());
+                continue;
+            }
+            SortedMap<Integer, Map<String, String[]>> objectMap = result.get(className);
+            if (objectMap == null) {
+                objectMap = new TreeMap<>();
+                result.put(className, objectMap);
+            }
+            // Get the property from the right object #objectNumber of type 'objectName'; create it if they don't exist
+            Map<String, String[]> object = objectMap.get(classNumber);
+            if (objectMap.get(classNumber) == null) {
+                object = new HashMap<>();
+                objectMap.put(classNumber, object);
+            }
+            object.put(classPropertyName, parameter.getValue());
+        }
+        return result;
+    }
+
+    /**
+     * Create and/or update objects in a document given a list of HTTP parameters of the form
+     * '<spacename>.<classname>_<number>_<propertyname>'. If the object already exists, the field is replace by the
+     * given value. If the object doesn't exist in the document, it is created then the property <propertyname> is
+     * initialized with the given value. An object is only created if the given '<number>' is 'one-more' than the
+     * existing number of objects. For example, if the document already has 2 objects of type 'Space.Class', then it
+     * will create a new object only with 'Space.Class_2_prop=something'. Every other parameter like
+     * 'Space.Class_42_prop=foobar' for example, will be ignore.
+     * 
+     * @param eform Form information that contains all the query parameters
+     * @param context
+     * @throws XWikiException
+     * @since 7.0RC1
+     */
+    public void readObjectsFromFormUpdateOrCreate(EditForm eform, XWikiContext context) throws XWikiException
+    {
+        Map<String, SortedMap<Integer, Map<String, String[]>>> fromRequest =
+            parseRequestUpdateOrCreate(eform.getRequest(), context);
+        for (Entry<String, SortedMap<Integer, Map<String, String[]>>> requestClassEntries : fromRequest.entrySet()) {
+            DocumentReference requestClassReference =
+                getCurrentDocumentReferenceResolver().resolve(requestClassEntries.getKey());
+            SortedMap<Integer, Map<String, String[]>> requestObjectMap = requestClassEntries.getValue();
+            List<BaseObject> newObjects = new ArrayList<>();
+            for (Entry<Integer, Map<String, String[]>> requestObjectEntry : requestObjectMap.entrySet()) {
+                Integer requestObjectNumber = requestObjectEntry.getKey();
+                Map<String, String[]> requestObjectPropertyMap = requestObjectEntry.getValue();
+                BaseObject oldObject = getXObject(requestClassReference, requestObjectNumber);
+                if (oldObject == null) {
+                    // Create the object only if it has been numbered one more than the number of existing objects
```

Comment:
Not entirely relevant to this particular issue, since what you want to do now is update and/or create, but what about delete?

We currently have a separate action for deleting objects AFAIK, but if we are updating and creating in the same request, why not delete as well?

Basically, the case would be `oldObject == null && (requestObjectPropertyMap == null || requestObjectPropertyMap.isEmpty())`, signaling that the requests explicitly states that the object with a particular number is to be deleted, probably something like `space.className_objNumber=<emptyString>`.

Of course, this would need changing the parsing as well and is probably for a separate issue to be handled, if ever, and is obviously opened to discussion since some people might consider it error prone.